### PR TITLE
feat: use archive nodes to query attestations

### PR DESCRIPTION
**Pull request by SweeXordious**
## Overview

Supports querying for historical data when getting attestations that are pruned.

Would allow deploying the contract and also relay attestations if the corresponding valset got pruned via getting it from an archive node.

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

## Summary by CodeRabbit

- **New Features**
  - Enhanced querying capabilities with new functions to retrieve historical attestations and valsets by nonce and height.
  - Introduced a fallback mechanism in the `ProcessAttestation` function to recover from failed valset queries.
  - Added new flags `FlagCoreRPCHost` and `FlagCoreRPCPort` to specify the RPC address host and port.

- **Refactor**
  - Renamed the variable `querier` to `appQuerier` for clarity.
  - Modified the `getStartingValset` function to use the `tmQuerier` and `appQuerier` instances.

- **Tests**
  - Added new test functions to validate the new query functions and fallback mechanism.

- **Bug Fixes**
  - Added error handling in `getStartingValset` to handle cases where the valset cannot be retrieved from the node state.

### CODE REVIEW

**Comment by SweeXordious** [comment]
@coderabbitai review

File: cmd/blobstream/deploy/cmd.go [hunk 1]
```diff
@@ -43,19 +43,31 @@ func Command() *cobra.Command {
...
+				err := tmQuerier.Stop()
+				if err != nil {
+					logger.Error(err.Error())
+				}
+			}(tmQuerier)
+
+			vs, err := getStartingValset(cmd.Context(), *tmQuerier, appQuerier, config.startingNonce)
 			if err != nil {
 				logger.Error("couldn't get valset from state (probably pruned). connect to an archive node to be able to deploy the contract")
 				return errors.Wrap(
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply cmd/blobstream/deploy/cmd.go]: [0dec09b](https://github.com/celestiaorg/orchestrator-relayer/pull/571/commits/0dec09b4c5f727b9bc565237ed18aee45f312c2d)

File: rpc/app_querier.go [hunk 2]
```diff
@@ -170,6 +287,43 @@
...
+			&celestiatypes.QueryLatestValsetRequestBeforeNonceRequest{Nonce: nonce},
+			grpc.Header(&header),
+		)
+		if err == nil {
+			return resp.Valset, err
+		}
+		aq.Logger.Debug("keeping looking for attestation in archival state", "err", err.Error())
+		currentHeight -= uint64(BlocksIn20DaysPeriod)
+	}
+	return nil, ErrNotFound
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/app_querier.go]: e1d3468

File: rpc/app_querier.go [hunk 3]
```diff
@@ -153,6 +246,30 @@
...
+		latestValset, err := aq.QueryHistoricalLastValsetBeforeNonce(ctx, latestNonce, currentHeight)
+		if err == nil {
+			return latestValset, nil
+		}
+
+		aq.Logger.Debug("keeping looking for attestation in archival state", "err", err.Error())
+		currentHeight -= uint64(BlocksIn20DaysPeriod)
+	}
+	return nil, ErrNotFound
+}
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/app_querier.go]: e1d3468

File: rpc/app_querier.go [hunk 4]
```diff
@@ -63,6 +70,57 @@ func (aq *AppQuerier) QueryAttestationByNonce(ctx context.Context, nonce uint64)
...
+			if err != nil {
+				return nil, err
+			}
+			return unmarshalledAttestation, nil
+		}
+		aq.Logger.Debug("keeping looking for attestation in archival state", "err", err.Error())
+		currentHeight -= uint64(BlocksIn20DaysPeriod)
+	}
+	return nil, ErrNotFound
+}
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/app_querier.go]: e1d3468

File: rpc/app_querier.go [hunk 5]
```diff
@@ -170,6 +308,53 @@ func (aq *AppQuerier) QueryLastValsetBeforeNonce(ctx context.Context, nonce uint
...
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Valset, nil
+}
+
+// QueryRecursiveHistoricalLastValsetBeforeNonce recursively looks for the last historical valset before nonce for a certain height until genesis.
+func (aq *AppQuerier) QueryRecursiveHistoricalLastValsetBeforeNonce(ctx context.Context, nonce uint64, height uint64) (*celestiatypes.Valset, error) {
```
Comment (evan-forbes) [inline rpc/app_querier.go:328]: are both of these used independently, or do we only ever call one of these?

**Review by evan-forbes — COMMENTED** [review]
do all of these need to be expoesed by the api? also can we not use latest and last? similar to our other previous discussions around this I think it would be good to just use Last and earliest.
`QueryHistoricalAttestationByNonce`, `QueryRecursiveHistoricalAttestationByNonce`, `QueryHistoricalLatestAttestationNonce`, `QueryHistoricalValsetByNonce`, `QueryRecursiveLatestValset`, `QueryHistoricalLastValsetBeforeNonce`, `QueryRecursiveHistoricalLastValsetBeforeNonce`.

these names are very confusing imo

File: rpc/app_querier.go (hunk 5 above)
Comment (SweeXordious) [reply rpc/app_querier.go:328]: One of them queries at an exact height, the other goes range by range of blocks to find the attestation. So yes, both is used

**Comment by SweeXordious** [comment]
I prefer to do these renames:

> do all of these need to be expoesed by the api? also can we not use latest and last? similar to our other previous discussions around this I think it would be good to just use Last and earliest.

in a separate PR. Since that would require renaming different parts of the implementation and I don't want to include all of that in this PR.

**Review by evan-forbes — DISMISSED** [review]
the api increase isn't that big of deal since its just the orchestrator/relayer, and might require a larger PR to fix properly. This is just something that was sticking was all. Ideally I think it would be great to isolate the actual queries made to the app in the app querier, and then refactor and separate the logic for getting the last, earliest, before nonce / after nonce

it would be nice to be consistent with last and earliest like we are doing in the app, as I often get confused since I'm not 100% sure we're using latest consistently. I could also just not be understanding latest.

**Comment by SweeXordious** [comment]
@evan-forbes is ready for review again, only added a relayer test with an archive node.

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**